### PR TITLE
Commonize and combine various bits of Node functionality

### PR DESCRIPTION
**Pull request by daviddrysdale**
# Checklist

- [x] Pull request affects core Oak functionality (e.g. runtime, SDK, ABI)
  - [x] I have written tests that cover the code changes.
  - [x] I have checked that these tests are run by [Cloudbuild](cloudbuild.yaml)
  - [ ] I have updated [documentation](docs/) accordingly.
  - [ ] I have raised an [issue](https://github.com/project-oak/oak/issues) to
        cover any TODOs and/or unfinished work.
- [ ] Pull request includes prototype/experimental work that is under
      construction.

### CODE REVIEW

File: oak/server/rust/oak_runtime/src/node/grpc/server.rs [hunk 1]
```diff
@@ -133,16 +120,16 @@ impl GrpcServerNode {
...
     }
+}
 
-    /// Main Node worker thread.
-    fn worker_thread(mut self) {
-        let pretty_name = pretty_name_for_thread(&thread::current());
-
+/// Oak Node implementation for the gRPC server.
+impl Node for GrpcServerNode {
+    fn run(mut self: Box<Self>) {
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/node/grpc/server.rs]: Not sure why `mut` is needed here?

File: oak/server/rust/oak_runtime/src/runtime/mod.rs [hunk 2]
```diff
@@ -1195,13 +1191,16 @@ impl Runtime {
...
-            }
-        };
+        let runtime = self.clone();
+        let node_join_handle = thread::Builder::new()
+            .name(node_name.to_string())
+            .spawn(move || {
+                node_instance.run();
+                // It's now safe to remove the state for this Node, as there's nothing left
+                // that can invoke `Runtime` functionality for it.
+                runtime.remove_node_id(node_id)
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/runtime/mod.rs:1197]: It seems a bit odd to have this removal logic here, I guess it's fine, but I wonder if it should be somewhere else (e.g. just in the runtime termination logic). Did you consider alternatives?

File: oak/server/rust/oak_runtime/src/node/grpc/server.rs [hunk 3]
```diff
@@ -60,38 +51,31 @@ impl GrpcServerNode {
...
-        let read_status = self
-            .runtime
-            .wait_on_channels(&[self.initial_reader])
-            .map_err(|error| {
-                error!("Couldn't wait on the initial reader handle: {:?}", error);
-                OakStatus::ErrInternal
-            })?;
+    fn init_channel_writer(
+        &mut self,
+        runtime: &RuntimeProxy,
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/node/grpc/server.rs:61]: Not sure I fully agree with the changes in commit 1d20c1bc62b708c296c35c87bde125d13f307b66 . As much as I like functional idioms, this seems to introduce an artificial distinction between the runtime proxy and all the other fields. I mean, you can take this to the extreme and have an empty struct and pass everything as function args, but it kind of defeats the point of encapsulation.

File: oak/server/rust/oak_runtime/src/runtime/mod.rs [hunk 4]
```diff
@@ -618,26 +606,35 @@ impl Runtime {
...
-            .drain()
-        {
-            debug!("join thread for {:?}...", node_id);
-            if let Err(err) = instance.join() {
-                error!("could not terminate node: {:?}", err);
+        let join_handles = self.take_join_handles();
+        for (node_id, join_handle_opt) in join_handles {
+            if let Some(join_handle) = join_handle_opt {
+                debug!("join thread for {:?} ...", node_id);
+                if let Err(err) = join_handle.join() {
```
Comment (ipetr0v) [inline oak/server/rust/oak_runtime/src/runtime/mod.rs:613]: Will it stop a Node that is currently blocked on waiting on the TCP socket?

File: oak/server/rust/oak_runtime/src/node/wasm/mod.rs [hunk 5]
```diff
@@ -744,11 +744,15 @@ impl WasmNode {
         }
     }
 
-    fn validate_entrypoint(&self) -> Result<(), OakStatus> {
-        let abi = WasmInterface::new(self.config_name.clone(), self.runtime.clone());
+    fn validate_entrypoint(
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply oak/server/rust/oak_runtime/src/node/wasm/mod.rs]: Done.

File: oak/server/rust/oak_runtime/src/node/wasm/mod.rs [hunk 6]
```diff
@@ -744,11 +744,15 @@ impl WasmNode {
...
+        runtime: RuntimeProxy,
+        module: Arc<wasmi::Module>,
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply oak/server/rust/oak_runtime/src/node/wasm/mod.rs]: Done.

File: oak/server/rust/oak_runtime/src/runtime/mod.rs (hunk 2 above)
Comment (daviddrysdale) [reply oak/server/rust/oak_runtime/src/runtime/mod.rs:1199]: TBH, having the node removal logic anywhere *other* than here seems odd to me.  This is the point where the Node's thread completes, so having it anywhere would require gratuitous inter-thread synchronization, and deferring it to runtime termination would mean that `node_infos` has unbounded growth.

File: oak/server/rust/oak_runtime/src/node/grpc/server.rs (hunk 3 above)
Comment (daviddrysdale) [reply oak/server/rust/oak_runtime/src/node/grpc/server.rs:61]: As well as merging four different copies of the same thing (pass proxy+handle at construction, store in per-Node-type struct field, make use of those field (only) in `run()`), I also think of it as tightening up the scope of the things available to the Node: if the Node is not currently running, then it should not need access to a `Runtime` instance, nor be able to do any messaging using a Handle.

File: oak/server/rust/oak_runtime/src/node/grpc/server.rs [hunk 7]
```diff
@@ -37,74 +34,41 @@ use tonic::{
 };
 
 /// Struct that represents a gRPC server pseudo-Node.
-///
-/// For each gRPC request from a client, gRPC server pseudo-Node creates a pair of temporary
-/// channels (to write a request to and to read a response from) and passes corresponding handles to
-/// the [`GrpcServerNode::channel_writer`].
```
[the remark this answers is not in the document]
Comment (daviddrysdale) [reply oak/server/rust/oak_runtime/src/node/grpc/server.rs:43]: Yes, but I've got another PR in preparation that does a commenting update across the runtime.

File: oak/server/rust/oak_runtime/src/runtime/mod.rs (hunk 4 above)
Comment (daviddrysdale) [reply oak/server/rust/oak_runtime/src/runtime/mod.rs:615]: I don't know, but I don't think this PR would change that behaviour.  Or is there some existing trigger that I've missed?

Comment (ipetr0v) [reply oak/server/rust/oak_runtime/src/runtime/mod.rs:615]: No, I think everything will be the same

File: oak/server/rust/oak_runtime/src/node/grpc/server.rs (hunk 1 above)
Comment (daviddrysdale) [reply oak/server/rust/oak_runtime/src/node/grpc/server.rs]: At this point the call to `self.init_channel_writer` below requires a `mut self`.  This gets removed in a later commit though.

File: oak/server/rust/oak_runtime/src/node/mod.rs [hunk 8]
```diff
@@ -117,47 +107,50 @@ pub fn check_port(address: &SocketAddr) -> Result<(), ConfigurationError> {
...
+    }
+
+    pub fn node_subname(&self, entrypoint: &str) -> String {
+        match self {
+            Configuration::LogNode => "LogNode".to_string(),
+            Configuration::GrpcServerNode {
+                address: _,
+                tls_identity: _,
+            } => "GrpcServerNode".to_string(),
+            Configuration::WasmNode { module: _ } => format!("WasmNode-{}", entrypoint),
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/node/mod.rs:152]: Why is `module` captured and discarded? And why does it have an underscore as type? It seems you should be able to just ignore all the fields with `...`. Similarly elsewhere.

File: oak/server/rust/oak_runtime/src/node/mod.rs [hunk 9]
```diff
@@ -117,47 +107,50 @@ pub fn check_port(address: &SocketAddr) -> Result<(), ConfigurationError> {
...
-                initial_handle,
-            )),
-            Configuration::External => Box::new(external::PseudoNode::new(
-                config_name,
-                runtime,
-                initial_handle,
-            )),
+            ))),
+            Configuration::WasmNode { module } => {
+                match wasm::WasmNode::new(node_name, module.clone(), entrypoint) {
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/node/mod.rs:134]: (alternative)

```rust
wasm::WasmNode::new(node_name, module.clone(), entrypoint)
  .map(|node| Box::new(node))
```

File: oak/server/rust/oak_runtime/src/node/wasm/mod.rs [hunk 10]
```diff
@@ -91,9 +89,9 @@ impl WasmInterface {
     }
 
     /// Creates a new `WasmInterface` structure.
-    pub fn new(pretty_name: String, runtime: RuntimeProxy) -> WasmInterface {
+    pub fn new(pretty_name: &str, runtime: Option<RuntimeProxy>) -> WasmInterface {
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/node/wasm/mod.rs:92]: Can you clarify in the comment why the second arg is an `Option`?

File: oak/server/rust/oak_runtime/src/node/wasm/mod.rs [hunk 11]
```diff
@@ -718,74 +716,73 @@ impl wasmi::ModuleImportResolver for WasiStub {
     }
 }
 
+fn validate_entrypoint(module: &wasmi::Module, entrypoint: &str) -> Result<(), OakStatus> {
+    let abi = WasmInterface::new("validate_entrypoint", None);
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/node/wasm/mod.rs:720]: This seems quite hacky... Is there a nicer way that the validation logic can be encapsulated in something that can be reused more ergonomically? If not now, perhaps add a TODO?

File: oak/server/rust/oak_runtime/src/node/mod.rs (hunk 8 above)
Comment (daviddrysdale) [reply oak/server/rust/oak_runtime/src/node/mod.rs]: I forgot about that syntax &ndash; done.

File: oak/server/rust/oak_runtime/src/node/mod.rs (hunk 9 above)
Comment (daviddrysdale) [reply oak/server/rust/oak_runtime/src/node/mod.rs:134]: I tried that but it doesn't work (some kind of failure to convert between concrete type `wasm::WasmNode` and trait type`dyn node::Node`).

File: oak/server/rust/oak_runtime/src/node/wasm/mod.rs (hunk 10 above)
Comment (daviddrysdale) [reply oak/server/rust/oak_runtime/src/node/wasm/mod.rs]: Went for a different approach, see below.

File: oak/server/rust/oak_runtime/src/node/wasm/mod.rs (hunk 11 above)
Comment (daviddrysdale) [reply oak/server/rust/oak_runtime/src/node/wasm/mod.rs]: I thought about an alternative that was less hacky but I thought would involve more boilerplate/copied-stuff.  However, now I come to try it out, it's not too bad &ndash; it adds a new Wasm host function resolver just for validation (which therefore doesn't need any state), and by pushing the body of the `WasmInterface::resolv_func` method into a standalone function, the copy-pasta I was worried about doesn't happen. WDYT?

File: oak/server/rust/oak_runtime/src/runtime/mod.rs (hunk 4 above)
Comment (tiziano88) [reply oak/server/rust/oak_runtime/src/runtime/mod.rs:615]: AFAICT `join` just waits for the thread to naturally terminate, it does not really signal anything at all, the other thread would not even know it is been waited on.

File: oak/server/rust/oak_runtime/src/node/wasm/mod.rs [hunk 12]
```diff
@@ -537,113 +530,134 @@ impl wasmi::Externals for WasmInterface {
...
-                        ValueType::I64, // handle
-                    ][..],
-                    Some(ValueType::I32),
-                ),
+        oak_resolve_func(field_name, signature)
+    }
+}
+
+/// A resolver function, mapping Oak host function names to an index and a type signature.
+fn oak_resolve_func( field_name: &str, signature: &wasmi::Signature, ) -> Result<wasmi::FuncRef, wasmi::Error> {
```
Comment (tiziano88) [inline oak/server/rust/oak_runtime/src/node/wasm/mod.rs:543]: Not sure what's going on with the formatting.

Comment (daviddrysdale) [reply oak/server/rust/oak_runtime/src/node/wasm/mod.rs]: Fixed (something up with my auto-format save-hook)

**Comment by tiziano88** [comment]
*I think* this PR now causes the runtime tests to not actually run the node body under test, and therefore assertions do not have any effect:

https://github.com/project-oak/oak/blob/master/oak/server/rust/oak_runtime/src/runtime/tests.rs

To reproduce, add a test case similar to the following (which is worth keeping checked in):

```rust
/// Checks that a panic in the node body actually causes the test case to fail, and does not
/// accidentally get ignored.
#[test]
#[should_panic]
fn panic_check() {
    run_node_body(
        Label::public_trusted(),
        Box::new(|_runtime| {
            panic!("testing that panic works");
        }),
    );
}
```

I think the solution is to wait for the termination of the node before exiting the test case.

**Comment by daviddrysdale** [comment]
Oops, yes, the test body now needs to synchronize completion (and also stopping the runtime would make sense).

**Comment by daviddrysdale** [comment]
(BTW, the panic test may not be a reliable detector, because of #905)

**Comment by tiziano88** [comment]
It's not as bad as I thought anyways, if there is a failing assertion / panic, it will still fail, though after the test case has finished running (which makes for a confusing error message, but at least it does not get completely ignored)

**Comment by tiziano88** [comment]
The panic test seemed to work when I tried locally, but I am not sure whether it is reliable indeed.

**Pull request merged**